src/py21cmfast/wrapper.py

```diff
@@ -3219,3 +3223,308 @@ def calibrate_photon_cons(
+    cosmo_params : :class:`~CosmoParams`, optional
+        Defines the cosmological parameters used to compute initial conditions.
+    PARALLEL_APPROX : bool, optional
+        Flag for parrallel approximation, if True, parallel approximation is taken which is much quicker, but more approximate. Default: False.
+    rotation : bool, optional
+        Flag for rotation of boxes, if True boxes are shifted for every HII_DIM, which is the size of the simulation. Default: True.
+
+    Returns
+    -------
+    kSZ_output :
+        KSZ_box: map of the kSZ effect, in Kelvins.
+        taue_boxp: map of the optical depth.
+    """
+    if lc is None:
+        if user_params is None:
+            user_params = UserParams(**UserParams._defaults_)
+        if cosmo_params is None:
+            cosmo_params = CosmoParams(**CosmoParams._defaults_)
+        lc_quantities = ("brightness_temp", "xH_box", "density", "velocity_z")
+        lc = run_lightcone(
+            redshift=z_start,
+            user_params=user_params,
+            cosmo_params=cosmo_params,
+            lightcone_quantities=lc_quantities,
+            random_seed=random_seed,
+        )
+    random.seed(random_seed)
+
+    kSZ_consts = KszConstants(
+        lc.user_params.HII_DIM,
+        lc.user_params.BOX_LEN,
+        lc.cosmo_params.hlittle,
+        lc.cosmo_params.OMb,
+        len(lc.lightcone_redshifts),
+        z_start,
+        lc.lightcone_distances[0],
+        0.245,  # Helium fraction
+        3,  # redshift of helium reionization, for tau_e calculation
+    )
+
+    kSZ_consts.mean_taue_curr_z = _tau_e(
+        0, z_start, None, None, 0, cosmo_params, kSZ_consts
+    )  # mean optical depth
+
+    Tcmb, mean_taue_fin = _Proj_array(
+        lc.lightcone_redshifts,
+        lc.density,
+        lc.velocity,
+        lc.xH_box,
+        kSZ_consts,
+        cosmo_params,
+        PARALLEL_APPROX=PARALLEL_APPROX,
+        rotation=rotation,
+    )
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        P_k, l_s, err = get_power(
+            Tcmb
+            * KszConstants.CMperMPC
+            / constants.c.cgs.value
+            * 1e6
+            * cosmo_params.cosmo.Tcmb0.value
+            / np.sqrt(2 * np.pi),
+            lc.user_params.BOX_LEN,
+            bins=30,
+            log_bins=True,
+            get_variance=True,
+        )  # in microK^2
+    P_k = P_k * l_s ** 2
+    err = np.sqrt(err) * l_s ** 2
+    l_s *= lc.lightcone_distances[0]
+    return kSZ_output(
+        Tcmb
+        * KszConstants.CMperMPC
+        / constants.c.cgs.value
+        * cosmo_params.cosmo.Tcmb0.value,
+        mean_taue_fin,
+        l_s=l_s[np.logical_not(np.isnan(l_s))],
+        kSZ_power=P_k[np.logical_not(np.isnan(l_s))],
+        err=err[np.logical_not(np.isnan(l_s))],
+    )
+
+
+def _Proj_array(
+    redshifts,
+    density,
+    velocity,
+    xH,
+    kSZ_consts,
+    cosmo_params,
```

Comment:
not used, but again. defining all things in one big function allows you to not pass these as arguments